core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3459,7 +3459,7 @@ private void registerOperandSubQueries(
       final BigDecimal noTrailingZeros = bd.stripTrailingZeros();
       // If we don't strip trailing zeros we may reject values such as 1.000....0.
 
-      final int maxPrecision = typeSystem.getMaxNumericPrecision();
+      final int maxPrecision = typeSystem.getMaxPrecision(SqlTypeName.DECIMAL);
```

Comment:
Is this change related to the current problem? What problems would occur if we don't make any changes here?

Reply:
You can take a look at UBIGINT in SqlTypeName, which belongs to DECIMAL type family.
I also add some tests to make it more readable.

Reply:
> UBIGINT should not belong to the DECIMAL family. If it does, it's a bug.

My bad, it belongs to NUMERIC, and I think that's right. We use DECIMAL for UBIGINT's jdbc type.